jep/0000/README.adoc

```diff
@@ -0,0 +1,220 @@
+
+To follow-up with this specific sample, JNLPLauncher do define (at time writing) two deprecated constructors
+and another non-deprecated one :
+```java
+@Deprecated
+public JNLPLauncher(String tunnel, String vmargs) { ... }
+
+@Deprecated
+public JNLPLauncher() { ... }
+
+public JNLPLauncher(boolean enableWorkDir) { ... }
+
+```
+The sole one we can guarantee is designed to submit configuration is the @DataBound one. For others
+there's no deterministic way to choose one. Even considering the configuration attributes we have from
+yaml file, which one should we use if configuration doesn't define `enableWorkDir`, `tunnel` nor `vmargs`?
+first one passing `null, null` of non-deprecated one passing `false` default value ?
+
+=== During Jenkins development
+
+Actually the exact same code change discussed in https://github.com/jenkinsci/configuration-as-code-plugin/issues/233[issue 233]
+could have been implemented without breaking Configuration-as-Code inferred model :
+
+- by defining a `@DataBoundSetter` for newly introduced `workDirSettings` attribute
+- by not defining `workDirSettings` argument as `@Nonnull` and applying default value
+
+Assuming Configuration-as-Code would exist for some time when this change was applied to jenkins-core, we
+might have detected it during code review and prevented this to happen.
+
+Configuration-as-Code tooling wants to avoid such unnecessary breaking changes to happen in the future.
+But as it wasn't there before developers aren't aware of this constraint, and we can't just ask them
+to keep our use-case in mind any time they write some code (there's already enough backward compatibility
+issues for them to address).
+
+Configuration-as-Code tooling do define tooling to detect such breaking changes, and warn developer about
+them. This tooling do run the model inference logic and compare with previous state of the codebase. This
+will warn developer about breaking change being introduced for Configuration-as-Code users.
+
+Sometime such changes are actually required. For such situation we propose some metadata being added aside
+changes made to codebase so Configuration-as-Code can give end-user some hints on the break and how to fix
+it.
+
+=== Before applying configuration
+
+On end-user side we want ot offer a validation toolkit to be used when applying configuration on a jenkins
+instance. This toolkit do parse the yaml configuration, try to apply configuration on the inferred model,
+and will report issue detected during this dry-run configuration process.
+
+It would rely on breaking changes metadata to offer end-user some guidance on way to address repoerted
+backward compatibility issues.
+
+
+== Backwards Compatibility
+
+N/A
+
+== Security
+
+N/A
+
+== Infrastructure Requirements
+
+If approved, Configuration-as-Code tooling to check for API backward compatibility will need to be
+included in the CI process, so it can inform plugin developers and contributors on breaking changes
```

Comment:
Meaning what, exactly? That an incompatible change to databinding produces a build failure? Even assuming the other pieces have been developed, this is not trivial as you need to determine a baseline to compare against, which presumably means looking up the most recent Maven release older than the current snapshot.

Reply:
You need not necessary “store an inferred model between builds”. You can have a build compute the old model almost as easily as it computes the new model. But you need some way to deterministically identify the “old version”.